apps/hotel/src/domain/GuestStay.ts

```diff
@@ -0,0 +1,201 @@
+      if (state.checkedInAt) return []
+      return [Events.CheckedIn({ at })]
+    }
+
+  export const charge =
+    (at: OffsetDateTime, chargeId: ChargeId, amount: number): Decision =>
+    (state) => {
+      if (state.type !== "Active") throw new Error("Invalid charge")
+      if (state.charges.includes(chargeId)) return []
+      return [Events.Charged({ chargeId, at, amount })]
+    }
+
+  export const payment =
+    (at: OffsetDateTime, paymentId: PaymentId, amount: number): Decision =>
+    (state) => {
+      if (state.type !== "Active") throw new Error("Invalid payment")
+      if (state.payments.includes(paymentId)) return []
+      return [Events.Paid({ paymentId, at, amount })]
+    }
+
+  type CheckoutResult =
+    | { type: "Ok" }
+    | { type: "AlreadyCheckedOut" }
+    | { type: "BalanceOutstanding"; balance: number }
+
+  export const checkOut =
+    (at: OffsetDateTime): DecisionResult<CheckoutResult> =>
+    (state) => {
+      switch (state.type) {
+        case "Closed":
+          return [{ type: "Ok" }, []]
+        case "TransferredToGroup":
+          return [{ type: "AlreadyCheckedOut" }, []]
+        case "Active":
+          return state.balance === 0
+            ? [{ type: "Ok" }, [Events.CheckedOut({ at })]]
+            : [{ type: "BalanceOutstanding", balance: state.balance }, []]
+      }
+    }
+
+  type GroupCheckoutResult = { type: "Ok"; residualBalance: number } | { type: "AlreadyCheckedOut" }
+
+  export const groupCheckout =
+    (at: OffsetDateTime, groupId: GroupCheckoutId): DecisionResult<GroupCheckoutResult> =>
+    (state) => {
+      switch (state.type) {
+        case "Closed":
+          return [{ type: "AlreadyCheckedOut" }, []]
+        case "TransferredToGroup":
+          if (state.groupId === groupId) {
+            return [{ type: "Ok", residualBalance: state.residualBalance }, []]
+          }
+          return [{ type: "AlreadyCheckedOut" }, []]
+        case "Active":
+          return [
+            { type: "Ok", residualBalance: state.balance },
+            [Events.TransferredToGroup({ at, groupId, residualBalance: state.balance })],
+          ]
+      }
+    }
+}
+
+export class Service {
+  constructor(private readonly resolve: (id: GuestStayId) => Decider<Events.Event, Fold.State>) {}
+
+  charge(id: GuestStayId, chargeId: ChargeId, amount: number, at = OffsetDateTime.now()) {
+    const decider = this.resolve(id)
+    return decider.transact(Decide.charge(at, chargeId, amount))
+  }
+
+  pay(id: GuestStayId, paymentId: PaymentId, amount: number, at = OffsetDateTime.now()) {
+    const decider = this.resolve(id)
+    return decider.transact(Decide.payment(at, paymentId, amount))
+  }
+
+  checkOut(id: GuestStayId, at = OffsetDateTime.now()) {
+    const decider = this.resolve(id)
+    return decider.transactResult(Decide.checkOut(at))
+  }
+
+  // driven exclusively by GroupCheckout
+  groupCheckout(id: GuestStayId, groupId: GroupCheckoutId, at = OffsetDateTime.now()) {
```

Comment:
not sure how this name arose and/or how it meshes with the comment in the source
but maybe react or some other better name can convey that more naturally

Reply:
it's your name! I'm just porting 😓

Reply:
yes, was waiting for that, but I know you will fix it in both!